### ISSUE TITLE
Open Source frameworks need a useable Oracle XE CI Image

### ISSUE THREAD

**Issue by deusaquilus** [oracle/docker-images#1156]
I am a developer of an open source framework that is desperately trying to support Oracle DB despite multiple roadblocks that Oracle has placed in front of me. Over the past 24 hours, you have issued DMCA takedowns of multiple images used by open source frameworks that are trying to support your database, so that your actual clients who pay you, can use your database in production environments with their OS tech stacks. For example, Lightbend's official image of an FRM framework called Slick relies on an image that you just took down: [slick/slick#1993](https://github.com/slick/slick/issues/1993). I am trying to build Oracle support in a similar framework that is also based on an image that you removed: [sath89/oracle-12c](https://hub.docker.com/r/sath89/oracle-12c). I've started a discussion of how to proceed here: getquill/quill#1295.

My question is the following:
 * Is there any possibility that we can have a 18c docker image that will not require a DockerHub login? 
Or:
 * Is there any possibility that can we can have an official 12c image that does not require a DockerHub login? 

Oracle's current 12c docker image is unuseable to Open Source frameworks using public CIs because:

1) It requires a private DockerHub account.
2) Private DockerHub credentials are impossible to use in public CIs (e.g. Travis) for any external contributor pull request (see [here](https://docs.travis-ci.com/user/pull-requests#pull-requests-and-security-restrictions) and [here](https://github.com/docker/hub-feedback/issues/1419)) because this is an security-impasse.
3) As a result of this, most Open Source frameworks have opted to use 3rd party DockerHub images which you are now taking down.

We really, really want to be able to support Oracle database in our products but we need a way to test them without having to enter private credentials. If DockerHub is a impossibility, maybe the solution is to have some kind of private Oracle Docker repo that only Travis/CircleCI can have access to. 
Can there be some sort of dialogue about this???

**Comment by gvenzl** [comment]
Hi @deusaquilus, thanks for notifying us about this. I have started investigating as to why the 11g XE image was taken down. Note, the 12c image was in violation with the license agreement that states that it is not allowed to redistribute the software.

We are exploring the possibility of having official XE images on Docker Hub going forward. So far there is no decision nor conclusion on that yet.

We approached Travis CI over a year ago wanting to make Oracle Database available to them for exactly that purpose. Unfortunately we have never heard back. We are more than happy to revisit that or perhaps with CircleCI. Do you happen to have any relationship with any of these CI companies so that we can start such a dialogue?

**Comment by deusaquilus** [comment]
Hi @gvenzl, thanks for looking into this issue! Any official 18c XE DockerHub image would work for me so long as it does not require DockerHub credentials. I really hope this becomes a reality!

**Comment by Djelibeybi** [comment]
I have changed the title of this issue as the admins of this GitHub repo have no control (or influence) over any legal action taken by Oracle, including the issuing of DMCA requests. I'd rather we focus on trying to get frameworks the images they need instead.

**Comment by deusaquilus** [comment]
@Djelibeybi It's a overly-sensational title, I agree. I'm just really hoping that all the open-source frameworks relying on 12c+ features have somewhere to go once the containers they rely on are taken down. I understand that "redistributing binaries" is a legal issue with entitles Oracle to do DMCAs but it would be a really big improvement on external developer-relations to have a big message somewhere saying: "here's something you can realistically use instead."

**Comment by fkorotkov** [comment]
Hey @gvenzl,

CTO of [Cirrus CI](https://github.com/marketplace/cirrus-ci) here. Would love to start a dialog with you about adding Oracle DBs support out of the box to Cirrus CI without even a need for an authentication into a private registry. Should fit pretty well into [additional containers](https://cirrus-ci.org/guide/writing-tasks/#additional-containers) feature. 👌 

If interested, shoot an email at fedor@cirruslabs.org. 🙌

**Comment by BanzaiMan** [comment]
Hi, @gvenzl,

> We approached Travis CI over a year ago wanting to make Oracle Database available to them for exactly that purpose. Unfortunately we have never heard back.

I am sorry about this bad experience. I searched for the record, but haven't found one yet. If you are up for it, we would love to discuss this more. Please email us at support@travis-ci.com.

Again, I apologize for the earlier lack of response.

We look forward to hearing from you.

**Comment by gvenzl** [comment]
Thanks @fkorotkov and @BanzaiMan, I have just reached out to both of you individually via email.

**Comment by deusaquilus** [comment]
@gvenzl As far as I understand 18c XE allows binary redistribution. Is this correct? Can I post a 18c XE image on DockerHub without having to be concerned about a potential DMCA?

**Comment by deusaquilus** [comment]
@gvenzl and @Djelibeybi. I am asking this question again. According to what I am reading here:
https://blogs.oracle.com/imc/oracle-database-18c-xe-available-for-everyone (and from what the license agreement [here](https://www.oracle.com/technetwork/licenses/db18c-express-license-5137264.html) seems to be saying). Oracle 18c-xe binaries can be redistributed. If I upload a 18c-xe image to dockerhub based on fuzziebrain's image ([here](https://github.com/fuzziebrain/docker-oracle-xe)). Can I reasonably expect not to get DMCA'd???

**Comment by ttrahan** [comment]
@gvenzl Head of BD here for CircleCI. We're also happy to have this discussion with you. Please reach out to me at tomt@circleci.com to get started. Looking forward to it!

**Comment by Djelibeybi** [comment]
@deusaquilus we are not lawyers and cannot answer your question. You would need legal counsel to review the license and advise you accordingly, especially considering you do not control the hosting and cannot (as such) control who is capable of downloading the binaries. Therefore, you may run afoul of the export compliance restrictions.

**Comment by deusaquilus** [comment]
@Djelibeybi I'm sorry if this is coming out the wrong way. I am not looking for legal advice. As an open source contributor, I don't have the resources or time to pursue legal issues in this manner at all so this is completely beside the point. The question is simply whether I have the ability to support Oracle in my framework (with my highly limited time and resources) or not. I think that all open source developers face a similar question, and as an employee of Oracle, you probably have a much, much, much better understanding the nuances of this matter then most of us. This is not a complaint, this is plea for help. Any advice or input you can provide is highly appreciated.

**Comment by gvenzl** [comment]
Hi all,

Here some updates on this:

- CirrusCI has just merged a new feature to support private Docker registries such as container-registry.oracle.com. [Feature request](https://github.com/cirruslabs/cirrus-ci-docs/issues/163), [Documentation](https://cirrus-ci.org/guide/linux/#working-with-private-registries)
- We have reached out to TravisCI and are opening the discussions
- We will reach out to CircleCI this week

Thanks everybody for your support and understanding!

**Comment by gvenzl** [comment]
Hi @deusaquilus, that is my understanding as well and, as I have stated, I'm investigating as to why the 11g XE image was taken down. Unfortunately I do not have any answer on that just yet.

**Comment by kyrrem** [comment]
I really have a though time understanding why we cannot have official Oracle XE images for 11g and 18c on docker hub. It is such a low effort thing to fix, which would make life so much easier for any company that wants to have an open source strategy with parts of their products.

Right now for our company we are forced to either fork our codebase to have one "internal" version that supports Oracle and an external one that only supports other database providers. Is this really in Oracles best interest - seeing that docker is becoming more and more a defacto standard?

**Comment by Djelibeybi** [comment]
> @Djelibeybi I'm sorry if this is coming out the wrong way. I am not looking for legal advice. 

I understand your frustration and I wish I could provide the advice you're after. If we were discussing one of the open source licenses, I'd be the right person to comment (I'm one of the product managers for Oracle Linux), but as this is a limited-use database license, @gvenzl is the right person to provide the best advice.

**Comment by Zed-Kun** [comment]
Hi @gvenzl, is there any update on this issue ?

**Comment by gvenzl** [comment]
Hi @Zed-Kun and everyone else,

Yes, we I have a couple of updates, thinks are still work in progress right now:

* We had a conversation with TravisCI and they are evaluating whether the license terms of Oracle XE and Oracle Database in general are in favor with their business practices. We are awaiting response from them.
* We have reached out to CircleCI and are trying to setup a conversation on the same matter
* We are in the process of making Oracle XE available on [container-registry.oracle.com](https://container-registry.oracle.com) and Docker Store. While that will still require a `docker login` we hope that this will simplify the matter for many users.
* We are working with our own business practices team to evaluate whether Oracle XE can be made available without click-through license agreement, a prerequisite to have it on Docker Hub as Docker Hub does not enforce any kind of click-through.
* I'm still investigating why the 11g XE image has been taken down.

I would like to thank you all and once again for your support and understanding.

**Comment by mloskot** [comment]
@deusaquilus 
> I am a developer of an open source framework that is desperately trying to support
> Oracle DB despite multiple roadblocks that Oracle has placed in front of me.

I feel your pain, https://github.com/wnameless/docker-oracle-xe-11g/issues/118#issuecomment-470049862
After many years of struggle, I'm still hopeful Oracle will consider to flatten the hurdles for FOSS community.

@gvenzl 
> We have reached out to TravisCI and are opening the discussions
> ...
> We are in the process of making Oracle XE available on container-registry.oracle.com and
> Docker Store. While that will still require a docker login 

Wouldn't such requirement of login make it a hurdle for any use on Travis CI?

**Comment by mijohansen** [comment]
Just to put my support for @deusaquilus and any other developers trying to support Oracle in their software. I'm working with the same issues as our production environment is running on Oracle. But we have serious problems adapting "the oracle way" to our workflows. This pain is an additional cost to the licence costs that of course follows an enterprise product. I was also using sath89/oracle-12c earlier, now everything is harder.

Right now I have serious performance issues in running my tests and while developing. I have used a [third party docker image](https://hub.docker.com/r/cosmomill/alpine-oracle-xe/) and injected the binaries. And since this can't be redistributed in any way, we have to do this everytime we set up a new environment.

We have good support in our organisation using Oracle, but that just slightly weights out the cost of having to deal with this kind of legacy licensing. In a sense it's a kind of debt that we will have to pay at some point. I have been wondering it its too late for Oracle to fix this problem as the quality/support of the software options around the product is decreasing. I guess this is yet another story about "distribution" where an giant fall in their own ignorance.

As I just discovered this repo, I will try to get them working. Maybe they will fix the performance issues. But hopefully @gvenzl and the other brilliant engineers at Oracle will convince their legal company overlords to help developers world wide to adopt, use and choose their products. At the moment they are slowly digging their own grave forcing large cooperations world wide to upgrade to something that is compatible with the way we want to work.

**Comment by yeroc** [comment]
@gvenzl has there been any meaningful progress towards getting an Oracle 18c XE docker image published on Docker Hub?

**Comment by elhigu** [comment]
@yeroc knex has been using this quillbuilduser/oracle-18-xe image since oracle 11 was removed from hub.

**Comment by yeroc** [comment]
@elhigu Thanks for pointing that out.  I think there would still be a lot of value in Oracle publishing an official one.  It would certainly make it much more discoverable.

**Comment by vmassol** [comment]
We are in the same situation on the XWiki open source project (https://xwiki.org). We want to support the maximum number of databases, including the Oracle Database. But the fact that there's no docker image readily available for testing without logging in, makes it very hard. 

I hope that Oracle legal department can fix this. Many thanks to @gvenzl for trying to fix this!

I've checked the repo at https://container-registry.oracle.com/ and it seems the latest version found there is 12.1.0.2 from 2.6 years ago.

Thanks

**Comment by vmassol** [comment]
@gvenzl Thank you for being so helpful in this thread. You're providing us hope :) Let's hope it won't be too long before we can have oracle DB images for testing without click-through.

I noticed that all the discussions were about 11g XE or 18 XE.

What about version 19.3.0-se? Is SE supposed to be the same as the previous XE in term of license?

https://blogs.oracle.com/imc/oracle-database-18c-xe-available-for-everyone mentions only version 18. Is it the same for 19.3 SE?

On the XWiki open source project, we decided to support the latest Oracle DB and we need to run our automated docker-based tests on it (locally on dev machines but also on our CI on the various agent machines). In your opinion, could we upload an XWiki-custom image of it (we need a custom image to avoid the 45mn of container startup time for ex) on dockerhub while waiting for a more generic solution to be found by Oracle?

Thanks

**Comment by cjbj** [comment]
@gvenzl I guess this is closed by #1531, which installs XE from yum.oracle.com without needing a separate manual download or click-through.

**Comment by Djelibeybi** [comment]
@cjbj not yet. We're still working on getting this onto the Docker Hub so it doesn't even need to be built locally.

**Comment by vmassol** [comment]
@Djelibeybi cool. Glad to see that the work is still in progress :) Do you think you could comment on the message I posted just above, at https://github.com/oracle/docker-images/issues/1156#issuecomment-573071783 ? 

All the discussions so far seem to be about XE but what about SE (the new name for 19.x AFAICS)? Is there also work so that it's available for development on dockerhub?

Thanks a lot

**Comment by Djelibeybi** [comment]
@gvenzl is the Database guy, so he'll have to reply to that question. But I don't believe your interpretation is correct. I believe we have 20c coming out this year in all three editions.

**Comment by vmassol** [comment]
>  I believe we have 20c coming out this year in all three editions.

Ok, I was assuming that maybe SE was the new term for XE since I could't find any XE version for 19.3.x. You seem to be saying that 20.x will have SE + XE. I guess I'm a bit lost now :)

Thanks!
